In [1]:
import pandas as pd
import matplotlib.pyplot as plt
import numpy as np


In [2]:
df = pd.read_csv('vehiculos.csv', delimiter=',', encoding='unicode_escape')

In [4]:
df.describe()

Price         Year     Kilometer       Engine       Length  \
count    1874.000000  1874.000000  1.874000e+03  1874.000000  1874.000000   
mean    17182.790368  2016.713447  5.317814e+04  1682.829242  4281.512807   
std     24260.903507     3.138477  5.878833e+04   633.055721   436.220747   
min       490.000000  1988.000000  0.000000e+00   624.000000  3099.000000   
25%      5000.000000  2015.000000  2.801975e+04  1197.000000  3985.000000   
50%      8424.995000  2017.000000  4.879750e+04  1497.000000  4360.000000   
75%     19082.500000  2019.000000  7.100000e+04  1995.000000  4620.000000   
max    350000.000000  2022.000000  2.000000e+06  6592.000000  5569.000000   

             Width       Height        Seats         Tank  
count  1874.000000  1874.000000  1874.000000  1874.000000  
mean   1767.886339  1588.967983     5.295091    52.217343  
std     131.344883   134.627659     0.807008    15.167250  
min    1475.000000  1213.000000     2.000000    15.000000  
25%    1695.000000  1485.000000     5.000000    42.000000  
50%    1770.000000  1544.000000     5.000000    50.000000  
75%    1831.000000  1670.750000     5.000000    60.000000  
max    2220.000000  1995.000000     8.000000   105.000000

## Pregunta con respecto al número de sillas de los vehículos y su precio es FALSO que:

### En promedio, son más costosos lso vehículos de dos asientos (Verdadero)

In [20]:
df.groupby(df['Seats'])['Price'].mean()

Seats
2    118000.000000
4     48004.614872
5     15194.240406
6     24199.564783
7     22314.250486
8     10818.063548
Name: Price, dtype: float64

### Hay menor variación en los precios de los vehículos de 5 asientos que en los de 4 asientos (Verdadero)


In [21]:
df.groupby(df['Seats'])['Price'].std()

Seats
2    118438.169523
4     56240.263620
5     21014.927055
6     25562.365252
7     21076.909330
8      6083.378604
Name: Price, dtype: float64

### En esta base de datos no hay información de vehículos de 3 asientos (Verdadero)

In [22]:
df[df['Seats'] == 3]

Empty DataFrame
Columns: [Make, Model, Price, Year, Kilometer, Fuel, Transmission, Location, Color, Owner, Seller, Engine, Power, Torque, Drivetrain, Length, Width, Height, Seats, Tank]
Index: []

### De los vehículos de 7 asientos, el mayor porcentaje corresponde a las marcas Toyota, Mahindra y Land Rover (Falso)

In [25]:
df[df['Seats'] == 7]['Make'].value_counts(normalize=True)*100

Make
Toyota           31.174089
Mahindra         24.696356
Maruti Suzuki    11.740891
Land Rover        5.263158
Audi              5.263158
Ford              4.858300
Mercedes-Benz     4.453441
Tata              3.643725
Hyundai           2.834008
Ssangyong         1.214575
Mitsubishi        1.214575
Honda             1.214575
Isuzu             0.809717
Datsun            0.404858
Kia               0.404858
Volkswagen        0.404858
Volvo             0.404858
Name: proportion, dtype: float64

## De las siguientes, la afirmación FALSA es

### El 96% de los vehículos de marca Suzuki con transmisión manual tiene una longitud de 4.49 metros o menos (Verdadero)

In [26]:
df.head(5)

Make              Model   Price  Year  Kilometer    Fuel  \
0           Tata          Nano Base   490.0  2010      60000  Petrol   
1        Hyundai            i10 Era  1300.0  2009      59000  Petrol   
2        Hyundai         Santro GLS  1410.0  2009      56400  Petrol   
3        Hyundai    Santro GL (CNG)  1450.0  2009      72000     CNG   
4  Maruti Suzuki  Wagon R LXi Minor  1450.0  2009      39586  Petrol   

  Transmission    Location   Color   Owner      Seller  Engine  \
0       Manual     Gurgaon   White   First  Individual     624   
1       Manual      Kanpur  Silver  Second  Individual    1086   
2       Manual  Jamshedpur  Silver   Third  Individual    1086   
3       Manual      Kanpur  Silver  Second  Individual    1086   
4       Manual      Ranchi     Red   First  Individual    1061   

               Power            Torque Drivetrain  Length  Width  Height  \
0            35@5250           48@3000        RWD    3099   1495    1652   
1            67@5500           99@2800        FWD    3565   1595    1550   
2  62 bhp @ 5500 rpm  96 Nm @ 3000 rpm        FWD    3565   1525    1590   
3  62 bhp @ 5500 rpm  96 Nm @ 3000 rpm        FWD    3565   1525    1590   
4            67@6200           84@3500        FWD    3520   1475    1690   

   Seats  Tank  
0      4  15.0  
1      5  35.0  
2      5  35.0  
3      5  35.0  
4      5  35.0

In [34]:
df[(df['Make'] == 'Maruti Suzuki') & (df['Transmission'] == 'Manual')]['Length'].quantile(0.96)

4490.0

### El 5% de los vehículos de marca Honda de color blanco tiene un motor de 1198 centímetros cúbicos o más. (Falso)

In [37]:
df[(df['Make'] == 'Honda') & (df['Color'] == 'White')]['Engine'].quantile(0.95)

1498.0

### Sólo el 1% de los vehículos con un tanque de 35 litros y del año 2015 o más recientes tienen una altura superior a los 1.7 metros (Verdadero)

In [38]:
df[(df['Tank'] == 35) & (df['Year'] >= 2015)]['Height'].quantile(0.99)

1700.0

### De la lista el carro más largo es un Rolls-Royce blanco o azul del año 2012 o 2011 (Verdadero)

In [43]:
df.sort_values(by=['Length'], ascending=False)[['Make', 'Color', 'Year', 'Length']].head(2)

Make  Color  Year  Length
1869  Rolls-Royce  White  2012    5569
1868  Rolls-Royce   Blue  2011    5569

## Para los vehículos de primer y segundo dueño es FALSO que:

### El 25% de los vehículos de primer dueño tienen más de 67.000 kilómetros (Verdadero)

In [44]:
df.head(2)

Make      Model   Price  Year  Kilometer    Fuel Transmission Location  \
0     Tata  Nano Base   490.0  2010      60000  Petrol       Manual  Gurgaon   
1  Hyundai    i10 Era  1300.0  2009      59000  Petrol       Manual   Kanpur   

    Color   Owner      Seller  Engine    Power   Torque Drivetrain  Length  \
0   White   First  Individual     624  35@5250  48@3000        RWD    3099   
1  Silver  Second  Individual    1086  67@5500  99@2800        FWD    3565   

   Width  Height  Seats  Tank  
0   1495    1652      4  15.0  
1   1595    1550      5  35.0

In [45]:
df[df['Owner'] == 'First']['Kilometer'].quantile(0.75)

67000.0

### Sería inusual que un automóvil de primer dueño tenga un kilometraje superior a los 129.690,6 kilómetros aprox (Verdadero)
Se necesita calcular el límite superior e inferior para los valores atípicos, en primer lugar se va a calcular el primer cuartil junto con el tercero.

También sale el rango intercuartil IQR

In [49]:
q1P3, q3P3 = df[df['Owner'] == 'First']['Kilometer'].quantile([0.25, 0.75])
IQR = q3P3 - q1P3

Ahora se van a sacar los límites inferior y superior

In [50]:
limInferior = q1P3 - 1.5*IQR
limSuperior = q3P3 + 1.5*IQR
print('Límite inferior:', limInferior, '\nLímite superior:', limSuperior)

Límite inferior: -37484.375 
Límite superior: 129690.625


### Hay 35 vehículos de primer dueño con un kilometraje inusualmente alto (Falso)

In [54]:
len(df[(df['Kilometer'] >= 129690) & (df['Owner'] == 'First')])

22

### Hay más vehículos con precio inusualmente alto en los de primer dueño que en los de segundo dueño (Verdadero)

In [55]:
len(df[(df['Kilometer'] >= 129690) & (df['Owner'] == 'First')]), len(df[(df['Kilometer'] >= 129690) & (df['Owner'] == 'Second')])

(22, 10)

## Con respecto a las estadísticas básicas de los automóviles es FALSO que

### El precio promedio de la totalidad de automóviles es de aprox. 17.182,8 (Verdadero)
### La mitad de los vehículos tiene un valor aprox. de 8425 dólares o menos (Verdadero)

In [56]:
df.describe()

Price         Year     Kilometer       Engine       Length  \
count    1874.000000  1874.000000  1.874000e+03  1874.000000  1874.000000   
mean    17182.790368  2016.713447  5.317814e+04  1682.829242  4281.512807   
std     24260.903507     3.138477  5.878833e+04   633.055721   436.220747   
min       490.000000  1988.000000  0.000000e+00   624.000000  3099.000000   
25%      5000.000000  2015.000000  2.801975e+04  1197.000000  3985.000000   
50%      8424.995000  2017.000000  4.879750e+04  1497.000000  4360.000000   
75%     19082.500000  2019.000000  7.100000e+04  1995.000000  4620.000000   
max    350000.000000  2022.000000  2.000000e+06  6592.000000  5569.000000   

             Width       Height        Seats         Tank  
count  1874.000000  1874.000000  1874.000000  1874.000000  
mean   1767.886339  1588.967983     5.295091    52.217343  
std     131.344883   134.627659     0.807008    15.167250  
min    1475.000000  1213.000000     2.000000    15.000000  
25%    1695.000000  1485.000000     5.000000    42.000000  
50%    1770.000000  1544.000000     5.000000    50.000000  
75%    1831.000000  1670.750000     5.000000    60.000000  
max    2220.000000  1995.000000     8.000000   105.000000

### Hay menor variabilidad en precios en los vehículos Mercedes-Benz que en los vehículos Honda (Falso)

In [68]:
df.loc[df['Make'] == 'Mercedes-Benz', 'Price'].std(), df.loc[df['Make'] == 'Honda', 'Price'].std()
    

(32929.857915990964, 2954.526426643345)

### Aproximadamente el 17.8% de los vehículos son marca Hyundai (Verdadero)

In [72]:
df['Make'].value_counts(normalize=True)*100

Make
Maruti Suzuki    21.237994
Hyundai          17.769477
Mercedes-Benz     8.484525
Honda             7.950907
Toyota            6.563501
Audi              6.296692
Mahindra          5.869797
BMW               4.749200
Tata              2.508004
Ford              2.508004
Renault           2.294557
Volkswagen        2.241195
Skoda             1.974386
Land Rover        1.600854
Kia               1.227321
Volvo             0.853789
Jeep              0.853789
Nissan            0.800427
MG                0.693703
Jaguar            0.693703
Porsche           0.640342
MINI              0.480256
Datsun            0.426894
Lexus             0.266809
Mitsubishi        0.213447
Chevrolet         0.213447
Rolls-Royce       0.160085
Ssangyong         0.160085
Isuzu             0.106724
Maserati          0.053362
Fiat              0.053362
Ferrari           0.053362
Name: proportion, dtype: float64

## De los vehículos de primer dueño con transmisión manual es FALSO que
### El precio promedio de los vehículos de primer dueño con transmisión manual es de aprox 7064.9 (Verdadero)

In [75]:
df[(df['Transmission'] == 'Manual') & (df['Owner'] == 'First')]['Price'].mean()

7064.900291715286

### El 95% de los precios de los vehículos de primer dueño con transmisión manual es 16500 dólares o menos (Verdadero)

In [77]:
df[(df['Owner'] == 'First') & (df['Transmission'] == 'Manual')]['Price'].quantile(0.95)

16500.0

### El coeficiente de variación del kilometraje de los vehículos de primer dueño con transmisión manual es de aprox 71% (Falso)

In [84]:
filtro = (df['Owner'] == 'First') & (df['Transmission'] == 'Manual')
df_filtrado = df[filtro]
cv_kilometraje = (df_filtrado['Kilometer'].std() / df_filtrado['Kilometer'].mean()) * 100

### Hay 17 marcas diferentes de vehículos primer dueño con transmisión manual (Verdadero)

In [82]:
len(df[(df['Transmission'] == 'Manual') & (df['Owner'] == 'First')]['Make'].value_counts())

17

# **Ayuda** <br>
1. Tabular una variable cualitativa X:
`df['X’].value_counts(normalize=False)`

2. Seleccionar solo las variables X, Y:
df[['X', 'Y']]

3. Filtrar cuando la variable X es x:
df[df['X']=='x']

4. Filtrar cuando la variable X=x & Y=y
df[(df['X']=='x') & (df['Y']='y') ]

5. Filtrar por texto de variable:
df[df['X'].str.contains('xxx')]

6. Ordenar por la variable X:
df.sort_values(by=['X'],ascending=True)

7. Bloquear por una o más condiones:
df.loc[(df['X'] == 'x') & (df['Y'] == 'y')]

8. Redondear decimales:
round(X,1)

9. Reemplazar un símbolo para corregir typos
df['X']=df['X'].str.replace('¥','Ñ')

10. Recodificar valores de variables
df['X'] = df['X'].replace({'a': 'A', 'b': 'B'})

11. Group by
df.groupby(['X','Y'])['Z'].aggregate(['count','mean','std’])

12. Crosstabulation
pd.crosstab(df['X'], df['Y'],normalize=False)

13. Crear una tabla dinámica
df.pivot_table(index='X',columns='Y',values='Z',aggfunc='mean')
